### PR TITLE
Fix negate_unsigned feature gate check

### LINKED ISSUE #24676: Unsigned negation in for inferred expressions doesn't warn
While this warns about unsigned negation:

```rust
let x : u8 = -1;
```

This does not:

```rust
let x = -1;
let y : u8 = x;
```

This especially egregious in places like range expressions where it can cause a loop to not run at all when the lower bound is negative (and therefore becomes a large unsigned integer).

### LINKED ISSUE #25206: 0u8..-1 compiles without warning
`0u8..-1` compiles, runs (as 0..255) without warning or error. Debug assertions don't fire.

We error on `0..-1u8` **error: unary negation of unsigned integers may be removed in the future**.

We can't "fix" this backwards compatibly after 1.0.

`0u8..-1` should either warn or trip debug assertions.

### LINKED ISSUE #26840: negate_unsigned feature gate triggered for user-defined Neg
```rust
struct S;
impl std::ops::Neg for S {
    type Output = u32;
    fn neg(self) -> u32 { 0 }
}
fn main() { -S; }
```
```
<anon>:6:13: 6:15 error: unary negation of unsigned integers may be removed in the future
<anon>:6 fn main() { -S; }
                     ^~
```

### PR DESCRIPTION
**Pull request by nagisa**
This commit fixes the negate_unsigned feature gate to appropriately
account for inferred variables.

This is technically a [breaking-change], but I’d consider it a bug fix.

cc @brson for your relnotes.

Fixes https://github.com/rust-lang/rust/issues/24676
Fixes #26840 
Fixes https://github.com/rust-lang/rust/issues/25206

### CODE REVIEW

**Comment by nagisa** [comment]
Note for reviewers: I put this into TypeLimits lint pass; I think negate_unsigned is strongly related to that, but if you want me to create a new pass I’ll do.

**Comment by eefriedman** [comment]
Umm, are you sure this patch does the right thing?  Specifically, it looks like the lint code only triggers for integer literals.

**Comment by nagisa** [comment]
Negation of unsigned variables is [another lint](https://github.com/nagisa/rust/blob/overflowing-unsigned/src/librustc_lint/builtin.rs#L153), if I understand what you’ve in mind.

**Comment by bluss** [comment]
Well it will be a very palpable breaking change, so just "technically breaking change" is misleading.

**Comment by eefriedman** [comment]
Feature gates are hard errors; lints are not (at least, https://github.com/nagisa/rust/blob/overflowing-unsigned/src/librustc_lint/builtin.rs#L153 is not).

**Comment by nagisa** [comment]
Yes. I put this feature gate in a lint, because otherwise the feature gate doesn’t quite work properly. Might need a comment about it or perhaps finding a better place for it.

**Comment by nagisa** [comment]
cc @pnkfelix, because you implemented the original gate.

**Comment by eefriedman** [comment]
Ignoring where exactly the code is, feature gate errors must be emitted using `emit_feature_err`.

Alternative approach: https://github.com/rust-lang/rust/compare/master...eefriedman:unary-typecheck .  Still WIP because I was having some trouble with the diagnostics, but it works otherwise.

**Comment by eefriedman** [comment]
Err, ignore me, I was making some bad assumptions; the current state of unsigned negation is really weird.  I need to do some research.

**Comment by nagisa** [comment]
> feature gate errors must be emitted using `emit_feature_err`

Isn’t that exactly what this PR does? My approach was to do it as simply as possible, though; if you find any better ways to fix the issue, great!

**Comment by eefriedman** [comment]
Ugh, I was just confusing myself; the current UNSIGNED_NEGATION lint is essentially dead.

Anyway, the problem is that the new code you inserted into the lint only runs for integer literals.

**Comment by nagisa** [comment]
> Anyway, the problem is that the new code you inserted into the lint only runs for integer literals.

Indeed, I was of impression that this is exactly how it should be, but having checked it now, currently feature gate actually works for unsigned variables. Will update!

**Comment by nagisa** [comment]
Updated!

Removed the dysfunctional lint as well. Tell me if you want it back for whatever reason.

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -89,12 +89,6 @@ impl LintPass for WhileTrue {
 }
 
 declare_lint! {
-    UNSIGNED_NEGATION,
-    Warn,
-    "using an unary minus operator on unsigned type"
-}
```
Comment (eefriedman) [inline src/librustc_lint/builtin.rs]: It would be nice to print a "deprecated lint" warning for existing code using `#![allow(unsigned_negation)]`, instead of an unknown_lint warning.

**Comment by nagisa** [comment]
Implemented lint deprecation in https://github.com/nagisa/rust/commit/22502154e6c1aee835554216d4b5054fd8f570f8.

**Comment by pnkfelix** [comment]
If I understand correctly, This PR goes straight from 1.0/1.1 accepting some code to now a hard error with an associated feature gate.

I think the current protocol for breaking changes is to first put in code that detects the scenario and emits a warning (potentially a hard warning, namely one that cannot be disabled), but not an error, at least not until the warning mode has made it to a release channel.

cc @rust-lang/lang

**Comment by nagisa** [comment]
> If I understand correctly, This PR goes straight from 1.0/1.1 accepting some code to now a hard error with an associated feature gate.

This is indeed the case. Due to very unfortunate implementation bugs, nor lint (added a while ago) nor the feature gate (added during 1.0 beta) fired for some cases of unsigned integer negation.

On the other hand, this PR also makes some valid code that was previously feature-gated (https://github.com/rust-lang/rust/issues/26840) actually valid.

> I think the current protocol for breaking changes is to first put in code that detects the scenario and emits a warning

This is easy to change (simple change from `emit_feature_err` to `span_warn`), but in that case I’m not overly enthusiastic on making sure feature gate still gates cases that were previously gated (that is, I’d propose to remove the gate altogether and emit a warning instead if we go this way).

**Comment by pnkfelix** [comment]
@nagisa to be clear: my only concern is that code that used to compile continues to compile for a cycle (potentially with a warning).  I do not think we care so much about ensuring that the feature gate continues to gate the same cases -- that is, it is okay for the compiler to become more liberal in what it accepts, as long as we are sure its headed in the right direction.  :)

**Comment by pnkfelix** [comment]
Actually on further reflection, I might be simply wrong to apply the aforementioned migration policy in this case.  In particular, this is not exactly analogous to the object lifetimes-defaults change, where we were concerned about the silent injection of new detaults causing hard-to-diagnose errors in downstream code -- in this case, it sounds like all of the cases where we inject an error should be pretty narrowly focused and thus easy for a user to understand the problem (and put in an appropriate fix)

**Comment by pnkfelix** [comment]
(The lang team discussed this last week but I forgot at that time to write a note)

The decision regarding policy was this: in general, assume that PR's with known breaking changes of the form "X becomes a compile-time error" simply have to go through a warning cycle before the hard error is turned on.

We did not discuss the detail about ensuring that the feature gate continues to feature gate the same things.  But my personal suspicion is that it is okay for the behavior there to change -- my main concern is about how we treat supposedly stable code.

For stable code, if we are correcting a previous oversight, then we are supposed to go through a warning cycle first.

So, in short, I think your "simple fix" from your [previous comment](https://github.com/rust-lang/rust/pull/27026#issuecomment-121944333) should be fine.
